core/src/main/java/io/grpc/internal/ManagedChannelImpl2.java

```diff
@@ -0,0 +1,874 @@
+      return oobChannel;
+    }
+
+    @Override
+    public String getAuthority() {
+      return ManagedChannelImpl2.this.authority();
+    }
+
+    @Override
+    public NameResolver.Factory getNameResolverFactory() {
+      return nameResolverFactory;
+    }
+
+    @Override
+    public void runSerialized(Runnable task) {
+      channelExecutor.executeLater(task).drain();
+    }
+
+    @Override
+    public void updatePicker(final SubchannelPicker picker) {
+      runSerialized(new Runnable() {
+          @Override
+          public void run() {
+            subchannelPicker = picker;
+            delayedTransport.reprocess(picker);
+          }
+        });
+    }
+  }
+
+  @Override
+  public LogId getLogId() {
+    return logId;
+  }
+
+  private class NameResolverListenerImpl implements NameResolver.Listener {
+    final LoadBalancer2 balancer;
+
+    NameResolverListenerImpl(LoadBalancer2 balancer) {
+      this.balancer = balancer;
+    }
+
+    @Override
+    public void onUpdate(final List<ResolvedServerInfoGroup> servers, final Attributes config) {
+      if (servers.isEmpty()) {
+        onError(Status.UNAVAILABLE.withDescription("NameResolver returned an empty list"));
+        return;
+      }
+      log.log(Level.FINE, "[{0}] resolved address: {1}, config={2}",
+          new Object[] {getLogId(), servers, config});
+      channelExecutor.executeLater(new Runnable() {
+          @Override
+          public void run() {
+            if (terminated) {
+              return;
+            }
+            try {
+              balancer.handleResolvedAddresses(servers, config);
+            } catch (Throwable e) {
+              log.log(Level.WARNING, "[" + getLogId() + "] Caught exception from LoadBalancer", e);
+              // It must be a bug! Push the exception back to LoadBalancer in the hope that it may
+              // be propagated to the application.
+              balancer.handleNameResolutionError(Status.INTERNAL.withCause(e)
+                  .withDescription("Thrown from handleResolvedAddresses(): " + e));
+            }
+          }
+        }).drain();
+    }
+
+    @Override
+    public void onError(final Status error) {
+      checkArgument(!error.isOk(), "the error status must not be OK");
+      log.log(Level.WARNING, "[{0}] Failed to resolve name. status={1}",
+          new Object[] {getLogId(), error});
+      channelExecutor.executeLater(new Runnable() {
+          @Override
+          public void run() {
+            if (terminated) {
+              return;
+            }
+            balancer.handleNameResolutionError(error);
+          }
+        }).drain();
+    }
+  }
+
+  private final class SubchannelImplImpl extends SubchannelImpl {
+    // Set right after SubchannelImplImpl is created.
+    InternalSubchannel subchannel;
+    final Object shutdownLock = new Object();
+    final Attributes attrs;
+
+    @GuardedBy("shutdownLock")
+    boolean shutdownRequested;
+    @GuardedBy("shutdownLock")
```

Comment:
we update `delayedShutdownTask` at line 839 without acquiring `shutdownLock`. should we remove this `@GuardedBy`?